external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala

```diff
@@ -83,6 +86,113 @@ private[kafka010] case class CachedKafkaConsumer private(
+    //      There is nothing to fetch, return null.
+    // scalastyle:on
+    if (offset >= untilOffset) {
+      // Case 2 or 8
+      // We seek to beginningOffset but beginningOffset >= untilOffset
+      reset()
+      return null
+    }
+
+    logDebug(s"Get $groupId $topicPartition nextOffset $nextOffsetInFetchedData requested $offset")
+    var outOfOffset = false
+    if (offset != nextOffsetInFetchedData) {
+      logInfo(s"Initial fetch for $topicPartition $offset")
+      seek(offset)
+      try {
+        poll(pollTimeoutMs)
+      } catch {
+        case e: OffsetOutOfRangeException =>
+          logWarning(s"Cannot fetch offset $offset, try to recover from the beginning offset", e)
+          outOfOffset = true
+      }
+    } else if (!fetchedData.hasNext()) {
+      // The last pre-fetched data has been drained.
+      seek(offset)
+      try {
+        poll(pollTimeoutMs)
+      } catch {
+        case e: OffsetOutOfRangeException =>
+          logWarning(s"Cannot fetch offset $offset, try to recover from the beginning offset", e)
+          outOfOffset = true
+      }
+    }
+    if (outOfOffset) {
+      val beginningOffset = getBeginningOffset()
+      if (beginningOffset <= offset) {
+        val latestOffset = getLatestOffset()
+        if (latestOffset <= offset) {
+          // Case 3 or 6
+          logWarning(s"Offset ${offset} is later than the latest offset $latestOffset. " +
+            s"Skipped [$offset, $untilOffset)")
+          reset()
+          return null
+        } else {
+          // Case 4 or 5
+          getAndIgnoreLostData(offset, untilOffset, pollTimeoutMs)
+        }
+      } else {
+        // Case 1 or 7
+        logWarning(s"Buffer miss for $groupId $topicPartition [$offset, $beginningOffset})")
+        return getAndIgnoreLostData(beginningOffset, untilOffset, pollTimeoutMs)
+      }
+    } else {
+      if (!fetchedData.hasNext()) {
+        // We cannot fetch anything after `polling`. Two possible cases:
+        // - `beginningOffset` is `offset` but there is nothing for `beginningOffset` right now.
+        // - Cannot fetch any date before timeout.
+        // Because there is no way to distinguish, just skip the rest offsets in the current
+        // partition.
+        logWarning(s"Buffer miss for $groupId $topicPartition [$offset, $untilOffset)")
+        reset()
+        return null
+      }
+
+      val record = fetchedData.next()
+      if (record.offset >= untilOffset) {
+        // Case 2
+        logWarning(s"Buffer miss for $groupId $topicPartition [$offset, $untilOffset)")
+        reset()
+        return null
+      } else {
+        if (record.offset != offset) {
+          // Case 1
+          logWarning(s"Buffer miss for $groupId $topicPartition [$offset, ${record.offset})")
+        }
+        nextOffsetInFetchedData = record.offset + 1
+        return record
+      }
+    }
+  }
+
+  private def reset(): Unit = {
+    nextOffsetInFetchedData = -2
```

Comment:
The use of -2 as a magic number here and elsewhere in the file is misleading, since the new consumer won't actually let you seek to -2 as an indicator of earliest.